### PR TITLE
pyk: initial KRun class

### PR DESCRIPTION
**Pull request by ehildenb**
This goal is to support krun-like interface too (execution from a single state), so we can build downstream frontends entirely in pyk:

- Rename and re-organize some test files (`*_test.py` for test helpers, `test_*.py` for tests).
- Makes the types of arguments in constructors `KPrint` and `KProve` stricter.
- Moves `use_directory` field from `KProve` to `KPrint`, and makes it a `Path` instead of a `str`.
- Renames field `main_file_name` of `KProve` to `main_file`, and makes it a `Path` instead of a `str`.
- Allows `_write_claim_definition` to work even if `KProve.main_file` is not defined.
- Adds new class `KRun`, and testing harness for it.
- Adds a testing harness for `KRun.run` in the integration tests.
- Factors out internal helper method `_kast` from `KPrint.{kast_to_kore,kore_to_kast}`.

### CODE REVIEW

File: pyk/src/pyk/ktool/kprint.py [hunk 1]
```diff
@@ -40,55 +40,63 @@
...
-            ntf.write(kore.text)
-            ntf.flush()
-            _LOGGER.info(f'Wrote file: {ntf.name}')
-            kast_command = ['kast', '--input', 'kore', '--output', 'json', ntf.name, '--definition', str(self.kompiled_directory)]
-            command_env = os.environ.copy()
-            proc_result = run_process(kast_command, _LOGGER, env=command_env)
-            if proc_result.returncode != 0:
-                raise RuntimeError(f'Calling kast failed: {kast_command}')
-            return KAst.from_dict(json.loads(proc_result.stdout)['term'])
+        output = _kast(self.definition_dir, kore.text, input='kore')
```
Comment (ehildenb) [inline pyk/src/pyk/ktool/kprint.py]:
```suggestion
        output = _kast(self.definition_dir, kore.text, input='kore', output='json')
```

File: pyk/src/pyk/ktool/krun.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+from .kprint import KPrint
+
+_LOGGER: Final = logging.getLogger(__name__)
+
+
+def _krun(
+    definition_dir: Path,
+    input_file: Path,
+    depth: Optional[int] = None,
+    args: List[str] = [],
```
Comment (nishantjr) [inline pyk/src/pyk/ktool/krun.py:20]: Rename this `additional_args` for consistency with `kompile`.

File: pyk/src/pyk/ktool/krun.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    main_module: str
+
+    def __init__(self, definition_dir: Path, use_directory: Optional[Path] = None) -> None:
+        super(KRun, self).__init__(definition_dir, use_directory=use_directory)
+        with open(self.definition_dir / 'backend.txt', 'r') as ba:
+            self.backend = ba.read()
+        with open(self.definition_dir / 'mainModule.txt', 'r') as mm:
+            self.main_module = mm.read()
+
+    def run(self, init_PGM: KInner, depth: Optional[int] = None, args: List[str] = []) -> CTerm:
```
Comment (nishantjr) [inline pyk/src/pyk/ktool/krun.py]:
```suggestion
    def run(self, init_PGM: KInner, depth: Optional[int] = None, additional_args: Iterable[str] = ()) -> CTerm:
```

Comment (nishantjr) [reply pyk/src/pyk/ktool/krun.py]: Its also dangerous using mutable values for defaults in python. See: https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments

File: pyk/src/pyk/integration_tests/krun_test.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    def tearDown(self):
+        shutil.rmtree(self.KRUN_USE_DIR, ignore_errors=True)
+        super().tearDown()
+
+    def assertTop(self, term):
+        self.assertEqual(term, mlTop())
+
+    def assertNotTop(self, term):
+        self.assertNotEqual(term, mlTop())
+
```
Comment (nishantjr) [inline pyk/src/pyk/integration_tests/krun_test.py]: These are duplicated in pyk/src/pyk/integration_tests/kprove_test.py . Lets move them to `kompiled_test.py`

**Review by nishantjr — APPROVED** [review]
Looks good. Some changes requested.

File: pyk/src/pyk/ktool/krun.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+    krun_command = ['krun', '--definition', str(definition_dir), str(input_file)]
+
+    if depth and depth >= 0:
+        args += ['--depth', str(depth)]
+
+    try:
+        return run_process(krun_command + args, _LOGGER)
+    except CalledProcessError as err:
+        raise RuntimeError(f'Command krun exited with code {err.returncode} for: {input_file}', err.stdout, err.stderr)
```
Comment (nishantjr) [inline pyk/src/pyk/ktool/krun.py]: Use excpetion chaining for better error messages.

```suggestion
        raise RuntimeError(f'Command krun exited with code {err.returncode} for: {input_file}', err.stdout, err.stderr) from err
```

File: pyk/src/pyk/ktool/krun.py (hunk 2 above)
Comment (ehildenb) [reply pyk/src/pyk/ktool/krun.py:20]: I'm actually renaming `kompile` to use `args` directly instead, which is more consistent with the other tools (`krun`, `kprove`, and `kprint`).

File: pyk/src/pyk/ktool/krun.py (hunk 3 above)
Comment (ehildenb) [reply pyk/src/pyk/ktool/krun.py]: Done.

File: pyk/src/pyk/integration_tests/krun_test.py (hunk 4 above)
Comment (ehildenb) [reply pyk/src/pyk/integration_tests/krun_test.py]: Done.

File: pyk/src/pyk/ktool/krun.py (hunk 5 above)
Comment (ehildenb) [reply pyk/src/pyk/ktool/krun.py]: Done.

File: pyk/src/pyk/ktool/krun.py (hunk 2 above)
Comment (ehildenb) [reply pyk/src/pyk/ktool/krun.py:20]: I'm going to go ahead and merge this, so I can open my next PR. Let me know if you want to see this switched back.

Comment (ehildenb) [reply pyk/src/pyk/ktool/krun.py:20]: Agh, I re-requested your review :P Not sure github will let me merge, but I'll leave automerge on for now.

**Pull request merged**